### PR TITLE
[FIX] l10n_mx: Fix account tags for COA SAT(XML) export

### PR DESCRIPTION
**Pull request by ZedarkMoon**
Steps to reproduce:
1. Install l10n_mx
2. Go to Accounting > Reporting > Trial balance.
3. Click the COA SAT(XML) export button.
4. Warnings are raised.

Problem: the report should work out-of-the-box on new databases.

Cause:
1. Some tags are missing on auto-generated accounts.
2. Some data were not correct and had the wrong tag. 102 accounts in particular should be debit and not credit.

This PR migrate the incorrect data and add a default naive computation of the tag on newly created accounts.

opw-3283746
Enterprise: https://github.com/odoo/enterprise/pull/57531

### CODE REVIEW

File: addons/l10n_mx/__init__.py [hunk 1]
```diff
@@ -2,3 +2,14 @@
...
 
 from . import models
+
+from odoo import api, SUPERUSER_ID
+
+
+def add_debit_tag(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    debit_accounts = env['account.account'].search([('code', '=like', '102%')])
+    debit_tag = env['account.account.tag'].search([('name', '=like', 'Debit Balance Account')])
```
Comment (aliyatastemirova) [inline addons/l10n_mx/__init__.py:12]: maybe we could ref by `tag_debit_balance_account` id instead, so it is more robust? :slightly_smiling_face:

File: addons/l10n_mx/__init__.py [hunk 2]
```diff
@@ -2,3 +2,14 @@
...
+    debit_accounts.write({
```
Comment (aliyatastemirova) [inline addons/l10n_mx/__init__.py:13]: `debit_accounts.tag_ids = debit_tag` should work

**Review by aliyatastemirova — CHANGES_REQUESTED** [review]
hi, thanks, just a few comments :slightly_smiling_face: 
also, I personally don't like adding internal odoo links in commit messages :slightly_smiling_face:

**Review by aliyatastemirova — APPROVED** [review]

File: addons/l10n_mx/__manifest__.py [hunk 3]
```diff
@@ -33,6 +33,7 @@
     "depends": [
         "account",
     ],
+    'post_init_hook': 'add_debit_tag',
```
Comment (smetl) [inline addons/l10n_mx/__manifest__.py:36]: Is the post_init_hook called when you update your module?

Comment (aliyatastemirova) [reply addons/l10n_mx/__manifest__.py:36]: ping @ZedarkMoon :slightly_smiling_face:

File: addons/l10n_mx/__init__.py [hunk 4]
```diff
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (clbr-odoo) [inline addons/l10n_mx/__init__.py:1]: I guess useless diff now that you don't modify the `__init__.py` file :wink:

File: addons/l10n_mx/migrations/2.1/post-migrate.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import api, SUPERUSER_ID
+
+from odoo.addons.account.models.chart_template import update_taxes_from_templates
+
+
+def migrate(cr, version):
+    update_taxes_from_templates(cr, 'l10n_mx.mx_coa')
```
Comment (clbr-odoo) [inline addons/l10n_mx/migrations/2.1/post-migrate.py:8]: Why updating taxes ? This method is only for updating tax's tags, I don't see the relation with this PR ? Am I missing something ?

File: addons/l10n_mx/migrations/2.1/post-migrate.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    debit_tag = env.ref("l10n_mx.tag_debit_balance_account")
+    credit_tag = env.ref("l10n_mx.tag_credit_balance_account")
+    mx_companies = env['res.company'].search([('country_id', '=', env.ref('base.mx').id)])
```
Comment (clbr-odoo) [inline addons/l10n_mx/migrations/2.1/post-migrate.py:13]: Search on companies that uses the chart template would be more robust.

File: addons/l10n_mx/migrations/2.1/post-migrate.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    # First update current entries
+    cr.execute("""
```
Comment (clbr-odoo) [inline addons/l10n_mx/migrations/2.1/post-migrate.py:16]: I don't really understand why you are doing all this in SQL, it feels too complicated for such a tiny change :thinking: 

What I would do:
- Either you go in pre-migrate and do it in SQL - but you can do something way more simple I think - 
- Either you stay in post-migrate and make use of the ORM's availability.

```python
    env = api.Environment(cr, SUPERUSER_ID, {})
    debit_tag = env.ref("l10n_mx.tag_debit_balance_account")
    credit_tag = env.ref("l10n_mx.tag_credit_balance_account")
    account_102_ids = env['ir.model.data'].search([
        ('name', 'ilike', '%_cuenta102_02'),
        ('model', '=', 'account.account'),
    ]).mapped('res_id')
    accounts_102 = env['account.account'].search([('id', 'in', account_102_ids)])
    accounts_102.tag_ids = [Command.unlink(credit_tag.id), Command.link(debit_tag.id)]
```
I think something like this does the trick ?

**Review by clbr-odoo — CHANGES_REQUESTED** [review]
Hello @ZedarkMoon !
Laurent asked me to have a look at this since I worked a lot on module migrations last year :slightly_smiling_face: 

I tried to reproduce the bug explained in the ticket but I think it's no longer happening, now it's only complaining that accounts automatically created by Odoo don't have tags, you can set them and export without error.
The change from credit -> debit is still relevant tho, but you'll need to update your commit message please.

File: addons/l10n_mx/migrations/2.1/post-migrate.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+from odoo import api, Command, SUPERUSER_ID
+
+
+def migrate(cr, version):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    debit_tag = env.ref("l10n_mx.tag_debit_balance_account")
+    credit_tag = env.ref("l10n_mx.tag_credit_balance_account")
+    mx_companies = env['res.company'].search([('account_fiscal_country_id', '=', env.ref('base.mx').id)])
+    accounts_102 = env['account.account'].search([('code', '=like', '102%'), ('company_id', 'in', mx_companies.ids)])
+    accounts_102.tag_ids = [Command.unlink(credit_tag.id), Command.link(debit_tag.id)]
```
Comment (clbr-odoo) [inline addons/l10n_mx/migrations/2.1/post-migrate.py:11]: I double-checked with Laurent, this is too risky to do: we could end up changing wrong 102 accounts. The users could have changed accounts.
The only safe way is to *only* migrate our data `cuenta102_02` and let the user handle his other 102 accounts by hand.
Generally speaking searching company by fiscal country is not safe either, better use the chart_template, but in this case I think you can just re-use the snippet I suggested last review ? :thinking:

**Review by clbr-odoo — CHANGES_REQUESTED** [review]
Sorry, few more things are not clear yet :slightly_smiling_face:

File: addons/l10n_mx/models/account_account.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        # EXTENDS account - ensure there is a tag on created MX accounts
+        # The computation is a bit naive and might not be correct in all cases.
+        accounts = super().create(vals_list)
+        debit_tag = self.env.ref('l10n_mx.tag_debit_balance_account')
+        credit_tag = self.env.ref('l10n_mx.tag_credit_balance_account')
+        mx_account_no_tags = accounts.filtered(lambda a: a.company_id.country_code == 'MX' and not a.tag_ids & (credit_tag + debit_tag))
+        DEBIT_CODES = ['1', '5', '6', '7']  # all other codes are considered "credit"
+        for account in mx_account_no_tags:
+            tag = debit_tag.id if account.code[0] in DEBIT_CODES else credit_tag.id
+            account.tag_ids = [Command.link(tag)]
```
Comment (smetl) [inline addons/l10n_mx/models/account_account.py:17]:
```suggestion
            tag_id = debit_tag.id if account.code[0] in DEBIT_CODES else credit_tag.id
            account.tag_ids = [Command.link(tag_id)]
```

**Comment by smetl** [comment]
@robodoo r+

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**